dll/cpl/sysdm/environment.c

```diff
@@ -17,6 +25,114 @@ typedef struct _VARIABLE_DATA
     LPTSTR lpCookedValue;
 } VARIABLE_DATA, *PVARIABLE_DATA;
 
+typedef struct _DIALOG_DATA
+{
+    BOOL bIsFriendlyUI;
+    BOOL bIsItemSelected;
+    DWORD dwSelectedValueIndex;
+    DWORD cxMin;
+    DWORD cyMin;
+    DWORD cxOld;
+    DWORD cyOld;
+    HWND hEditBox;
+    PVARIABLE_DATA VarData;
+} DIALOG_DATA, *PDIALOG_DATA;
+
+static DWORD
+GatherDataFromEditBox(HWND hwndDlg,
+                      PVARIABLE_DATA VarData)
+{
+    DWORD dwNameLength;
+    DWORD dwValueLength;
+
+    dwNameLength = SendDlgItemMessage(hwndDlg, IDC_VARIABLE_NAME, WM_GETTEXTLENGTH, 0, 0);
+    dwValueLength = SendDlgItemMessage(hwndDlg, IDC_VARIABLE_VALUE, WM_GETTEXTLENGTH, 0, 0);
+
+    if (dwNameLength == 0 || dwValueLength == 0)
+    {
+        return 0;
+    }
+    if (VarData->lpName == NULL)
+    {
+        VarData->lpName = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
+    }
+    else if (_tcslen(VarData->lpName) < dwNameLength)
+    {
+        GlobalFree(VarData->lpName);
+        VarData->lpName = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
+    }
```

Comment:
Seems this can be simplified (to just have one single GlobalAlloc) to:
```suggestion
    /* Reallocate the name buffer, regrowing it if necessary */
    if (!VarData->lpName || (_tcslen(VarData->lpName) < dwNameLength))
    {
        if (VarData->lpName)
            GlobalFree(VarData->lpName);
        VarData->lpName = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
    }
```